internal/jobcontainers/job.go

```diff
@@ -0,0 +1,308 @@
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to query for PIDs in job object: %s", err)
+	}
+
+	bufInfo := (*winapi.JOBOBJECT_BASIC_PROCESS_ID_LIST)(unsafe.Pointer(&buf[0]))
+	rawPids := make([]uintptr, bufInfo.NumberOfProcessIdsInList)
+
+	err = winapi.RtlMoveMemory(
+		(*byte)(unsafe.Pointer(&rawPids[0])),
+		(*byte)(unsafe.Pointer(&bufInfo.ProcessIdList[0])),
+		uintptr(bufInfo.NumberOfProcessIdsInList)*unsafe.Sizeof(rawPids[0]))
+	if err != nil {
+		return nil, fmt.Errorf("failed to move PID info to new buffer: %s", err)
+	}
+
+	pids := make([]uint32, bufInfo.NumberOfProcessIdsInList)
+	for i, rawPid := range rawPids {
+		pids[i] = uint32(rawPid)
+	}
+	return pids, nil
+}
+
+// Polls the IO completion port for notifications. Used for detecting when all of the
+// processes in a job have exited and for (TODO: dcantah) limit thresholds being reached.
+func (job *jobObject) pollIOCP() (uint32, error) {
+	var (
+		overlapped uintptr
+		qty        uint32
+		key        uint32
+	)
+	if job.iocpHandle != 0 {
+		if err := windows.GetQueuedCompletionStatus(job.iocpHandle, &qty, &key, (**windows.Overlapped)(unsafe.Pointer(&overlapped)), windows.INFINITE); err != nil {
+			return 0, err
+		}
+		return qty, nil
+	}
+	return 0, errors.New("IOCP handle is 0")
+}
+
+func (job *jobObject) queryMemoryStats() (*winapi.JOBOBJECT_MEMORY_USAGE_INFORMATION, error) {
+	info := winapi.JOBOBJECT_MEMORY_USAGE_INFORMATION{}
+	if err := winapi.QueryInformationJobObject(
+		job.jobHandle,
+		winapi.JobObjectMemoryUsageInformation,
+		uintptr(unsafe.Pointer(&info)),
+		uint32(unsafe.Sizeof(info)),
+		nil,
+	); err != nil {
+		return nil, fmt.Errorf("failed to query for job object memory stats: %s", err)
+	}
+	return &info, nil
+}
+
+func (job *jobObject) queryProcessorStats() (*winapi.JOBOBJECT_BASIC_ACCOUNTING_INFORMATION, error) {
+	info := winapi.JOBOBJECT_BASIC_ACCOUNTING_INFORMATION{}
+	if err := winapi.QueryInformationJobObject(
+		job.jobHandle,
+		winapi.JobObjectBasicAccountingInformation,
+		uintptr(unsafe.Pointer(&info)),
+		uint32(unsafe.Sizeof(info)),
+		nil,
+	); err != nil {
+		return nil, fmt.Errorf("failed to query for job object process stats: %s", err)
+	}
+	return &info, nil
+}
+
+func (job *jobObject) queryStorageStats() (*winapi.JOBOBJECT_BASIC_AND_IO_ACCOUNTING_INFORMATION, error) {
+	info := winapi.JOBOBJECT_BASIC_AND_IO_ACCOUNTING_INFORMATION{}
+	if err := winapi.QueryInformationJobObject(
+		job.jobHandle,
+		winapi.JobObjectBasicAndIoAccountingInformation,
+		uintptr(unsafe.Pointer(&info)),
+		uint32(unsafe.Sizeof(info)),
+		nil,
+	); err != nil {
+		return nil, fmt.Errorf("failed to query for job object storage stats: %s", err)
+	}
+	return &info, nil
+}
+
+// Assigns an IO completion port to get notified of events for the registered job
+// object.
+func attachIOCP(job windows.Handle, iocp windows.Handle) (int, error) {
+	info := winapi.JOBOBJECT_ASSOCIATE_COMPLETION_PORT{
+		CompletionKey:  uintptr(job),
+		CompletionPort: iocp,
+	}
```

Comment:
A struct like this would also need to be updated. I think (though again, I haven't checked so you'd need to do some testing) that you can just set the `CompletionKey` to `job` directly and that should still function properly.

Reply:
`job` is just `windows.Handle` which is just `uintptr` under the hood so yes. I did this for some reason I feel...